src/ua_securechannel.c

```diff
@@ -355,102 +876,492 @@ UA_SecureChannel_processSequenceNumber(UA_SecureChannel *channel, UA_UInt32 Sequ
+        return retval;
+
+    *requestId = sequenceHeader.requestId;
+
+    *bodySize = messageHeader->messageHeader.messageSize - *processedBytes;
 
-            /* Does the token match? */
-            if(tokenId != channel->securityToken.tokenId) {
-                if(tokenId != channel->nextSecurityToken.tokenId)
-                    return UA_STATUSCODE_BADCOMMUNICATIONERROR;
-                UA_SecureChannel_revolveTokens(channel);
+    return UA_STATUSCODE_GOOD;
+}
+
+/**
+ * \brief Processes an asymmetric chunk, decoding and decrypting it.
+ *
+ * \param chunk the chunk to process. The data in the chunk will be modified in place.
+                That is, it will be decoded and decrypted in place.
+ * \param channel the UA_SecureChannel to work on.
+ * \param messageHeader the message header of the chunk that was already decoded.
+ * \param processedBytes the already processed bytes. The offset must be relative to the start
+ *                       of the chunk. After this function finishes, the processedBytes offset
+ *                       will point to the beginning of the message body.
+ * \param requestId the requestId of the chunk. Will be filled by the function.
+ * \param bodySize the size of the body segment of the chunk.
+ */
+static UA_StatusCode
+UA_SecureChannel_processAsymmetricOPNChunk(const UA_ByteString* const chunk,
+                                           UA_SecureChannel* const channel,
+                                           UA_SecureConversationMessageHeader* const messageHeader,
+                                           size_t* const processedBytes,
+                                           UA_UInt32* const requestId,
+                                           size_t* const bodySize) {
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+
+    size_t chunkSize = messageHeader->messageHeader.messageSize;
+
+    UA_AsymmetricAlgorithmSecurityHeader clientAsymHeader;
+    UA_AsymmetricAlgorithmSecurityHeader_init(&clientAsymHeader);
+    retval |= UA_AsymmetricAlgorithmSecurityHeader_decodeBinary(chunk,
+                                                                processedBytes,
+                                                                &clientAsymHeader);
+    size_t messageAndSecurityHeaderOffset = *processedBytes;
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_AsymmetricAlgorithmSecurityHeader_deleteMembers(&clientAsymHeader);
+        return retval;
+    }
+
+    retval |= UA_AsymmetricAlgorithmSecurityHeader_copy(&clientAsymHeader,
+                                                        &channel->remoteAsymAlgSettings);
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_AsymmetricAlgorithmSecurityHeader_deleteMembers(&clientAsymHeader);
+        return retval;
+    }
+
+
+    if(channel->endpoint == NULL) {
+        // iterate available security policies and choose the correct one
+        UA_Endpoint *endpoint = NULL;
+
+        for(size_t i = 0; i < channel->endpoints->count; ++i) {
```

Comment:
Some really minor comments. This is really just cosmetics.
But if we can reduce complexity for the reader, we should.

- Reduce the max indendation in the loop by using `continue` when ByteString_equal fails.
- Use something like `UA_Endpoint *tmpEndpoint = &channel->endpoints->endpoints[i];` and write the actual endpoint only just before `break`. So you don't have to "pingpoing" back NULL.
- Use `if(!channel->endpoint)` to check for non-null.